readme.md
# BL808 IOT SDK

## Setup

1. Download sdk

    ```shell
    mkdir ~/bl808 && cd ~/bl808
    git clone https://github.com/bouffalolab/bl_iot_sdk.git
    git checkout bl808
    ```
    
2. Download toolchain [here](https://occ-oss-prod.oss-cn-hangzhou.aliyuncs.com/resource//1663142243961/Xuantie-900-gcc-elf-newlib-x86_64-V2.6.1-20220906.tar.gz) . Extract it to `toolchain/riscv/Thead_riscv/` and rename to  `Linux_x86_64` 

    ```shell
    cd ~/bl808 && mkdir -p bl_iot_sdk/toolchain/riscv/Thead_riscv
    
    mv {your_download_path}/Xuantie-900-gcc-elf-newlib-x86_64-V2.6.1-20220906.tar.gz ~/bl808
    tar -zxvf Xuantie-900-gcc-elf-newlib-x86_64-V2.6.1-20220906.tar.gz -C bl_iot_sdk/toolchain/riscv/Thead_riscv/
    
    cd bl_iot_sdk/toolchain/riscv/Thead_riscv && mv Xuantie-900-gcc-elf-newlib-x86_64-V2.6.1/ Linux_x86_64 && cd -
    ```

3. Build the uvc example

    ```shell
    # Compile e907 firmware
    cd ~/bl808/bl_iot_sdk/customer_app/bl808/bl808_demo_uvc_sd

    ./genromap

    # Compile c906 firmware
    cd ~/bl808/bl_iot_sdk/customer_app/bl808/bl808_demo_mm

    ./genromap
    ```

4. The compilation is complete, and the target file is saved in the `build_out` path

##  Flash Guide

1. Startup flash tool

    - Ubuntu
    ```shell
    cd ~/bl808/bl_iot_sdk/tools/flash_tool

    # startup
    ./BLDevCube-ubuntu &> /dev/null &
    ```
    - Windows

      execute `BLDevCube.exe`

2. Download bl808 firmware

    - Use the type-c usb cable to connect from the PC side to the UART port of the board

    - Startup `BLDevCube Tools` and choose `BL808` chip

    - Configure the tool according to the label below, finally click `Create & Download` will start to download firmware

        ![2022-12-19_15-39](tools/picture/2022-12-19_15-39.png)


3. About partition table files

    The partition table of bl808 is written in toml file format. You can specify the flash size used by the firmware by modifying the partition table.

    The default partition table is provided in the bl808/bl_iot_sdk/tools/flash_toolchips/bl808/partition/ directory, and you can also create your own partition table.

##  Start Run

1. press reset and you can get log from dual serial default baud rate `2000000`

     ![2022-12-19_16-07](tools/picture/2022-12-19_16-07.png)

2. Use the type-c usb connect tag ota to PC and run camera on PC you can get image from sensor

     ![2022-12-19_16-08](tools/picture/2022-12-19_16-08.png)


flash_tool/changelog.txt
= Release 1.6.6 =
Released: 2022-01-05

== Features ==
windows版本增加cklink烧写接口

== Changes ==
修改flash erase逻辑，防止erase过程中消耗过多时间
ubuntu版本主文件大小压缩到100MB以内
更新602/702的boot2和eflash loader bin
更新602/702支持的flash类型列表
更新602的mfg文件

== Bug Fixes ==
修复iap升级时设备枚举一直报错的问题
修复ubuntu下切换芯片时界面崩溃的问题
修复ubuntu下usb烧写出错的问题


= Release 1.6.5 =
Released: 2021-11-10

== Changes ==
更新BL702 MFG界面和固件
已经烧写过一次后在芯片没有复位还在运行eflash loader时，工具可在芯片不做任何动作的情况下继续烧写

== Bug Fixes ==
修复没有串口时无法生成image的bug
修复波特率填写非数字时烧写报错的bug
修复single和iap download的bug
修复jlink和openocd烧写出错的bug
修复BL602/702在使用加密功能时，efuse create会将key slot3同时读写保护的bug


= Release 1.6.4 =
Released: 2021-10-13

== Features ==
增加固件是否适配分区大小的检查

== Changes ==
BL602的MFG固件修改为2.43，支持efuse power offset ate的位置修改
修改702的eflash loader，支持根据flash efuse bit去烧录flash而不是自动scan，解决烧录过程中某些引脚高电平问题，同时支持指定引脚烧录

== Bug Fixes ==
修改FW_OTA.bin.hash生成的错误


= Release 1.6.3 =
Released: 2021-09-27

== Features ==
增加usb iap download功能


= Release 1.6.2 =
Released: 2021-09-03

== Bug Fixes ==
修复路径带空格的单文件烧写报错的bug


= Release 1.6.1 =
Released: 2021-09-01

== Features ==
boot2 bin文件增加flash配置，自动适配flash类型
eflash loader bin文件增加flash配置

== Changes ==
默认flash配置不再从bootheader中获取，而是在boot2中做适配，bootheader中的flash clock delay配置任然有效
boot2 iap烧写时，不再烧写efuse，boot2 iap烧写时，增大握手buff
文件对话框能自动记录上次打开的目录

== Bug Fixes ==
修复烧写过程中进度条出现的bug
修复MCU页面镜像类型选RAW时报错的bug


= Release 1.5.3 =
Released: 2021-06-04

== Features ==
MCU模式增加IAP页面
增加使用702作为usb转串口实现串口下载功能

== Bug Fixes ==
修复MacOS下无法显示串口的bug
修复dts文件解析字段长度被写死的bug
修复openocd无法自动关闭的bug

== Changes ==
使能IOT界面的Factory Params Checkbox


= Release 1.5.2 =
Released: 2021-04-15

== Features ==
增加Save Image按钮
增加File Serial协议

== Changes ==
串口显示窗口对PROG和SDIO串口进行标注
修改位置MCU界面Bind按钮位置
各芯片文件夹放至在chips目录下
各配置和程序文件放至utils目录下
切换页面时自动关闭串口监听


= Release 1.5.1 =
Released: 2021-03-31

== Changes ==
IOT页面删除board选择，增加dts文件选择
更新BL602的boot2

== Bug Fixes ==
修复出现特殊字符串日志保存异常的bug
修复read flash出错但提示成功的bug


= Release 1.5.0 =
Released: 2021-03-16

== Features ==
镜像生成增加RF的TLV信息
IOT页面增加ECC签名

== Changes ==
COM Port下拉菜单自动屏蔽虚拟串口
IOT生成的bin文件改放至img_create_iot文件夹下
MCU生成的bin文件改放至img_create_mcu文件夹下
修改BL602和BL702的RF MFG

== Bug Fixes ==
修复烧写文件路径带空格时报错的bug
修复单文件烧写报错的bug
修复jlink烧写报错的bug


= Release 1.4.8 =
Released: 2021-01-26

== Features ==
BL702支持加密烧写和OpenOCD烧写

== Changes ==
命令行选项用argparse替换getopt
加密库使用pycryptodome和cryptoplus


= Release 1.4.7 =
Released: 2020-12-21

== Features ==
BL702增加MFG工具
eflash loader增加日志记录

== Changes ==
美化界面显示
用yapf格式化代码
将修改后的fdt源码集成到lib里
更新img_create_cfg.conf文件
更新BL702的相关文件

== Bug Fixes ==
修复openocd的bug
修复打开uart的bug
修复标题栏图标不显示的bug
修复界面最大化和恢复时进度条显示的bug


= Release 1.4.6 =
Released: 2020-11-12

== Changes ==
修改校验和计算方法，提高烧写速度  #de63f77
减少img_load_process中的sleep时间  #9196b04
pack打包文件增加version.txt  #4e13639

== Bug Fixes ==
修改读取dts文件时报错的bug  #3a234cf 
修复在MacOS下界面选项无法显示的bug  #ee222b1


= Release 1.4.5 =
Released: 2020-11-05

== Changes ==
用csv库替代pandas
擦写超时从10000ms改为15000ms
eflash_loader_cfg.ini从conf拷贝时增加判断
eflash_loader.bin增加USB下载功能
修改pyinstaller打包脚本

== Bug Fixes ==
修复Bouffalo2Other.py不能转换mfg目录下的文件名的问题


= Release 1.4.4 =
Released: 2020-10-29

== Features ==
IOT增加efuse key自动加载

== Changes ==
更新mfg文件
更新eflash_loader文件

== Bug Fixes ==
修复转换TG版本时不能转换eflash_loader_cfg.conf中敏感信息的bug


= Release 1.4.3 =
Released: 2020-10-22

== Changes ==
天猫版本增加MCU界面
MFG页面的power增加10dbm选项
帮助菜单增加使用说明文档的链接
更新BL702的eflash_loader.bin
更新BL602的flash时钟类型

== Bug Fixes ==
修复在MacOS下烧写文件打包的bug


= Release 1.4.2 =
Released: 2020-10-14

== Changes ==
更新MFG固件
更新eflash loader
更新RF性能测试用户手册，增加产测说明

== Features ==
增加分区表overlay的判断
添加烧写文件打包功能
添加使用说明手册


= Release 1.4.1 =
Released: 2020-09-25

== Changes ==
将修改后的toml源码集成到lib里
修改定制版本的替换关键字

== Features ==
添加工具自动转换成定制版本的功能

== Bug Fixes ==
修复日志搜索框重复打开的bug


= Release 1.4.0 =
Released: 2020-09-16

== Changes ==
更新BL702文件

== Features ==
添加日志搜索功能

== Bug Fixes ==
修复toml解析路径的bug


= Release 1.3.8 =
Released: 2020-09-07

== Changes ==
更新mfg文件
用urllib替换requests
用pyaes替换Cryptodome
打包时用onedir模式替换onefile模式
将configobj集成到lib里

== Features ==
支持BL702芯片

== Bug Fixes ==
修复工具放大缩小时进度条显示问题


= Release 1.3.7 =
Released: 2020-08-14

== Changes ==
log日志显示页面增加时间显示
log日志显示页面打印语句标准化
更新boot2和mfg partition
eflash_loader增加xtal type选择
生成镜像时为阿里增加Hash
BL602的flash速度减少为40M

== Features ==
增加BL602烧写文件生成脚本

== Bug Fixes ==
修复Open Uart按钮打开串口后部分字符不能转换问题
修复MCU页面镜像地址错误仍然提示Create成功并继续烧写的错误


= Release 1.3.6 =
Released: 2020-08-05

== Changes ==
BL602/604和BL562/564的MCU界面Flash默认速度改为48M
删除对crc_ignore和hash_ignore的修改
更新MCU Flash Clock设置
更新BL602的eflash_loader.bin
更新eflash_loader_cfg.conf

== Features ==
增加自动/手动新版本并下载和更新的功能，提示框显示更新内容
MCU增加Flash Clock自动选择
MCU增加bind bootinfo/img函数

== Bug Fixes ==
修改选择BL60X和BL56X后Simple Flasher页面Xtal选项默认值为空的问题
修改Simple Flasher页面选择Enable后Factory Params不变灰的问题
修改eflash_loader的app_path问题



= Release 1.3.5 =
Released: 2020-07-27

== Changes ==
更新error code函数

== Features ==
增加日志窗口可浮动停靠功能
eflash loader增加auto download函数

== Bug Fixes ==
修复Image页面点击Clear按钮后进度条不清理的问题
修复bl60x生成镜像报错和烧写崩溃的bug
修复bl602镜像类型选择BLSP_Boot2时报错的bug
修复jlink下载报错的bug


= Release 1.3.4 =
Released: 2020-07-20

== Changes ==
Image页面Advanced选项修改成折叠面板的形式
更新bl602 eflash_loader.bin
更新eflash loader error code
选择烧写文件时默认处于上一次打开的文件夹

== Features ==
Flash Utils页面和Image页面增加Open URAT按钮，增加监听串口线程
增加日志文件保存，在log文件夹内自动生成
增加日志窗口拖放功能
增加日志窗口可浮动停靠功能

== Bug Fixes ==
修复在工具内切换芯片时tab页面闪烁的bug
修复flash erase选项的bug
修复下位机发送大量数据时工具卡死的问题
修复在Linux下运行出错的问题
修复bl60x下simple_flasher镜像生成的错误
修复Image页面点击Clear按钮后进度条不清理的问题


= Release 1.3.3 =
Released: 2020-07-10

== Changes ==
修改core内py文件文件名，将前缀bl60x改为bflb
修改工具名称为Bouffalo Lab Dev Cube，主文件名为BLDevCube
Partition页面根据芯片类型显示不同配置
Image页面Flash Clock选项HCLK改为BCLK

== Features ==
Flash Utils页面增加Erase Flash和Whole Chip选项
RF MFG页面BLE部分Power选项增加Auto选项

== Bug Fixes ==
修复Simple Flasher页面点击Board下拉框后不显示全路径的问题
修复bl60x芯片Image页面PLL无默认值的问题
修复切换芯片类型后MCU默认不在Image页面的问题
修复Simple Flasher页面AES加密没有对输入IV值检测的问题
修复Efuse Key页面Create按钮无效的问题


= Release 1.3.2 =
Released: 2020-07-06

== Changes ==
修改Simple Flasher页面布局和进度条信息显示功能
修改MCU下Image页面进度条信息显示功能
修改MCU下Image页面的PLL选项默认值为160M


image_conf/efuse_cfg_keys.py
# -*- coding:utf-8 -*-

efuse_cfg_keys = {
    "ef_sf_aes_mode": {
        "offset": "0",
        "pos": "0",
        "bitlen": "2"
    },
    "ef_sboot_sign_mode": {
        "offset": "0",
        "pos": "2",
        "bitlen": "2"
    },
    "ef_sboot_en": {
        "offset": "0",
        "pos": "4",
        "bitlen": "2"
    },
    "ef_cpu_enc_en": {
        "offset": "0",
        "pos": "7",
        "bitlen": "1"
    },
    "trim_enable": {
        "offset": "0",
        "pos": "12",
        "bitlen": "1"
    },
    "ef_sw_usage_1": {
        "offset": "0",
        "pos": "13",
        "bitlen": "1"
    },
    "ef_sdu_dis": {
        "offset": "0",
        "pos": "14",
        "bitlen": "1"
    },
    "ef_ble_dis": {
        "offset": "0",
        "pos": "15",
        "bitlen": "1"
    },
    "ef_wifi_dis": {
        "offset": "0",
        "pos": "16",
        "bitlen": "1"
    },
    "ef_0_key_enc_en": {
        "offset": "0",
        "pos": "17",
        "bitlen": "1"
    },
    "ef_sf_dis": {
        "offset": "0",
        "pos": "19",
        "bitlen": "1"
    },
    "ef_cpu_rst_dbg_dis": {
        "offset": "0",
        "pos": "21",
        "bitlen": "1"
    },
    "ef_se_dbg_dis": {
        "offset": "0",
        "pos": "22",
        "bitlen": "1"
    },
    "ef_efuse_dbg_dis": {
        "offset": "0",
        "pos": "23",
        "bitlen": "1"
    },
    "ef_dbg_jtag_dis": {
        "offset": "0",
        "pos": "26",
        "bitlen": "2"
    },
    "ef_dbg_mode": {
        "offset": "0",
        "pos": "28",
        "bitlen": "4"
    },
    "ef_dbg_pwd_low": {
        "offset": "4",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_dbg_pwd_high": {
        "offset": "8",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_ana_trim_0": {
        "offset": "12",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_sw_usage_0": {
        "offset": "16",
        "pos": "0",
        "bitlen": "32"
    },
    ###################################################################
    "bootrom_protect": {
        "offset": "16",
        "pos": "0",
        "bitlen": "1"
    },
    "uart_log_disable": {
        "offset": "16",
        "pos": "1",
        "bitlen": "1"
    },
    "sdio_pin_init": {
        "offset": "16",
        "pos": "2",
        "bitlen": "1"
    },
    "disable_cci_coexist": {
        "offset": "16",
        "pos": "3",
        "bitlen": "1"
    },
    "xtal_type": {
        "offset": "16",
        "pos": "4",
        "bitlen": "3"
    },
    "pll_clk": {
        "offset": "16",
        "pos": "7",
        "bitlen": "3"
    },
    "hclk_div": {
        "offset": "16",
        "pos": "10",
        "bitlen": "1"
    },
    "bclk_div": {
        "offset": "16",
        "pos": "11",
        "bitlen": "1"
    },
    "flash_clk_type": {
        "offset": "16",
        "pos": "12",
        "bitlen": "3"
    },
    "flash_clk_div": {
        "offset": "16",
        "pos": "15",
        "bitlen": "1"
    },
    "flash_cfg": {
        "offset": "16",
        "pos": "16",
        "bitlen": "2"
    },
    "flash_pwr_delay": {
        "offset": "16",
        "pos": "18",
        "bitlen": "2"
    },
    "tz_boot": {
        "offset": "16",
        "pos": "20",
        "bitlen": "1"
    },
    "encrypted_tz_boot": {
        "offset": "16",
        "pos": "21",
        "bitlen": "1"
    },
    "hbn_check_sign": {
        "offset": "16",
        "pos": "22",
        "bitlen": "1"
    },
    "keep_dbg_port_closed": {
        "offset": "16",
        "pos": "23",
        "bitlen": "1"
    },
    "mediaboot_disable": {
        "offset": "16",
        "pos": "24",
        "bitlen": "1"
    },
    "uartboot_disable": {
        "offset": "16",
        "pos": "25",
        "bitlen": "1"
    },
    "sdioboot_disable": {
        "offset": "16",
        "pos": "26",
        "bitlen": "1"
    },
    "hbn_jump_disable": {
        "offset": "16",
        "pos": "27",
        "bitlen": "1"
    },
    "jtag_switch": {
        "offset": "16",
        "pos": "28",
        "bitlen": "1"
    },
    "jtag_init": {
        "offset": "16",
        "pos": "29",
        "bitlen": "1"
    },
    "qfn40": {
        "offset": "16",
        "pos": "30",
        "bitlen": "1"
    },
    "debug_log_reopen": {
        "offset": "16",
        "pos": "31",
        "bitlen": "1"
    },
    ############################################## ###########################
    "ef_wifi_mac_low": {
        "offset": "20",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_wifi_mac_high": {
        "offset": "24",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_0_w0": {
        "offset": "28",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_0_w1": {
        "offset": "32",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_0_w2": {
        "offset": "36",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_0_w3": {
        "offset": "40",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_1_w0": {
        "offset": "44",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_1_w1": {
        "offset": "48",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_1_w2": {
        "offset": "52",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_1_w3": {
        "offset": "56",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_2_w0": {
        "offset": "60",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_2_w1": {
        "offset": "64",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_2_w2": {
        "offset": "68",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_2_w3": {
        "offset": "72",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_3_w0": {
        "offset": "76",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_3_w1": {
        "offset": "80",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_3_w2": {
        "offset": "84",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_3_w3": {
        "offset": "88",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_4_w0": {
        "offset": "92",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_4_w1": {
        "offset": "96",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_4_w2": {
        "offset": "100",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_4_w3": {
        "offset": "104",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_5_w0": {
        "offset": "108",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_5_w1": {
        "offset": "112",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_5_w2": {
        "offset": "116",
        "pos": "0",
        "bitlen": "32"
    },
    "ef_key_slot_5_w3": {
        "offset": "120",
        "pos": "0",
        "bitlen": "32"
    },
    "wr_lock_key_slot_4_l": {
        "offset": "124",
        "pos": "13",
        "bitlen": "1"
    },
    "wr_lock_key_slot_5_l": {
        "offset": "124",
        "pos": "14",
        "bitlen": "1"
    },
    "wr_lock_boot_mode": {
        "offset": "124",
        "pos": "15",
        "bitlen": "1"
    },
    "wr_lock_dbg_pwd": {
        "offset": "124",
        "pos": "16",
        "bitlen": "1"
    },
    "wr_lock_sw_usage_0": {
        "offset": "124",
        "pos": "17",
        "bitlen": "1"
    },
    "wr_lock_wifi_mac": {
        "offset": "124",
        "pos": "18",
        "bitlen": "1"
    },
    "wr_lock_key_slot_0": {
        "offset": "124",
        "pos": "19",
        "bitlen": "1"
    },
    "wr_lock_key_slot_1": {
        "offset": "124",
        "pos": "20",
        "bitlen": "1"
    },
    "wr_lock_key_slot_2": {
        "offset": "124",
        "pos": "21",
        "bitlen": "1"
    },
    "wr_lock_key_slot_3": {
        "offset": "124",
        "pos": "22",
        "bitlen": "1"
    },
    "wr_lock_key_slot_4_h": {
        "offset": "124",
        "pos": "23",
        "bitlen": "1"
    },
    "wr_lock_key_slot_5_h": {
        "offset": "124",
        "pos": "24",
        "bitlen": "1"
    },
    "rd_lock_dbg_pwd": {
        "offset": "124",
        "pos": "25",
        "bitlen": "1"
    },
    "rd_lock_key_slot_0": {
        "offset": "124",
        "pos": "26",
        "bitlen": "1"
    },
    "rd_lock_key_slot_1": {
        "offset": "124",
        "pos": "27",
        "bitlen": "1"
    },
    "rd_lock_key_slot_2": {
        "offset": "124",
        "pos": "28",
        "bitlen": "1"
    },
    "rd_lock_key_slot_3": {
        "offset": "124",
        "pos": "29",
        "bitlen": "1"
    },
    "rd_lock_key_slot_4": {
        "offset": "124",
        "pos": "30",
        "bitlen": "1"
    },
    "rd_lock_key_slot_5": {
        "offset": "124",
        "pos": "31",
        "bitlen": "1"
    },
}
